File: clients/swift/Sources/AiCup22/TcpStream.swift
#if os(Linux)
    import Glibc
#else
    import Darwin
#endif

class TcpStream : InputStream, OutputStream {
    init(_ host: String, _ port: Int) {
        #if os(Linux)
            sock = socket(AF_INET, Int32(SOCK_STREAM.rawValue), 0)
        #else
            sock = socket(AF_INET, SOCK_STREAM, 0)
        #endif
        if sock == -1 {
            fatalError("Failed to create socket")
        }
        var yes: Int32 = 1
        if setsockopt(sock, Int32(IPPROTO_TCP), TCP_NODELAY, &yes, socklen_t(MemoryLayout<Int32>.size)) < 0 {
            fatalError("Failed to set TCP_NODELAY")
        }
        var hints = addrinfo()
        var servinfo: UnsafeMutablePointer<addrinfo>?
        hints.ai_family = AF_INET
        #if os(Linux)
            hints.ai_socktype = Int32(SOCK_STREAM.rawValue)
        #else
            hints.ai_socktype = SOCK_STREAM
        #endif
        if getaddrinfo(host, String(port), &hints, &servinfo) != 0 {
            fatalError("Failed to get addr info")
        }
        if connect(sock, servinfo!.pointee.ai_addr, servinfo!.pointee.ai_addrlen) == -1 {
            fatalError("Failed to connect")
        }
        freeaddrinfo(servinfo)
    }

    deinit {
        if close(sock) != 0 {
            fatalError("Failed to close socket")
        }
    }

    func readBytesAtMost(_ byteCount: Int) -> [Byte] {
        var buffer = [Byte](repeating: 0x0, count: byteCount)
        let received = buffer.withUnsafeMutableBytes { recv(sock, $0.baseAddress, byteCount, 0) }
        if received < 0 {
            fatalError("Failed to read from socket")
        }
        return Array(buffer[..<received])
    }

    func writeBytes(_ data: [Byte]) {
        if send(sock, data, data.count, 0) < 0 {
            fatalError("Failed to write to socket")
        }
    }

    func flush() {
    }

    private let sock: Int32
}

File: clients/cpp_variant/TcpStream.cpp
#include "TcpStream.hpp"
#include <cstring>
#include <stdexcept>

TcpStream::TcpStream(const std::string& host, int port)
    : readBufferPos(0)
    , readBufferSize(0)
    , writeBufferPos(0)
    , writeBufferSize(0)
{
#ifdef _WIN32
    WSADATA wsa_data;
    if (WSAStartup(MAKEWORD(1, 1), &wsa_data) != 0) {
        throw std::runtime_error("Failed to initialize sockets");
    }
#endif
    sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock == -1) {
        throw std::runtime_error("Failed to create socket");
    }
    int yes = 1;
    if (setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, (char*)&yes, sizeof(int)) < 0) {
        throw std::runtime_error("Failed to set TCP_NODELAY");
    }
    addrinfo hints, *servinfo;
    std::memset(&hints, 0, sizeof(hints));
    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_STREAM;
    if (getaddrinfo(host.c_str(), std::to_string(port).c_str(), &hints,
            &servinfo)
        != 0) {
        throw std::runtime_error("Failed to get addr info");
    }
    if (connect(sock, servinfo->ai_addr, servinfo->ai_addrlen) == -1) {
        throw std::runtime_error("Failed to connect");
    }
    freeaddrinfo(servinfo);
}

void TcpStream::readBytes(char* buffer, size_t byteCount)
{
    while (byteCount > 0) {
        if (readBufferSize > 0) {
            if (readBufferSize >= byteCount) {
                memcpy(buffer, this->readBuffer + readBufferPos, byteCount);
                readBufferPos += byteCount;
                readBufferSize -= byteCount;
                return;
            }
            memcpy(buffer, this->readBuffer + readBufferPos, readBufferSize);
            buffer += readBufferSize;
            byteCount -= readBufferSize;
            readBufferPos += readBufferSize;
            readBufferSize = 0;
        }
        if (readBufferPos == BUFFER_CAPACITY) {
            readBufferPos = 0;
        }
        RECV_SEND_T received = recv(sock, this->readBuffer + readBufferPos + readBufferSize,
            BUFFER_CAPACITY - readBufferPos - readBufferSize, 0);
        if (received < 0) {
            throw std::runtime_error("Failed to read from socket");
        }
        readBufferSize += received;
    }
}

TcpStream::~TcpStream()
{
#ifdef _WIN32
    if (closesocket(sock) != 0)
#else
    if (close(sock) != 0)
#endif
    {
        throw std::runtime_error("Failed to close socket");
    }
}

void TcpStream::writeBytes(const char* buffer, size_t byteCount)
{
    while (byteCount > 0) {
        size_t capacity = BUFFER_CAPACITY - writeBufferPos - writeBufferSize;
        if (capacity >= byteCount) {
            memcpy(this->writeBuffer + writeBufferPos + writeBufferSize, buffer, byteCount);
            writeBufferSize += byteCount;
            return;
        }
        memcpy(this->writeBuffer + writeBufferPos + writeBufferSize, buffer, capacity);
        writeBufferSize += capacity;
        byteCount -= capacity;
        buffer += capacity;
        flush();
    }
}

void TcpStream::flush()
{
    while (writeBufferSize > 0) {
        RECV_SEND_T sent = send(sock, writeBuffer + writeBufferPos, writeBufferSize, 0);
        if (sent < 0) {
            throw std::runtime_error("Failed to write to socket");
        }
        writeBufferPos += sent;
        writeBufferSize -= sent;
    }
    writeBufferPos = 0;
}